AppArmor profile for rsync

Viktor-Yakovchuk opened the pull request:
Cover rsync calls from rsync package by the Apparmor profile

Fixed issue #10261

lxc-jenkins commented:
This pull request didn't trigger Jenkins as its author isn't in the allow list.

An organization member must perform one of the following:

 - To have this branch tested by Jenkins, use the "ok to test" command.
 - To have a one time test done, use the "test this please" command.

Those commands are simple Github comments of the format: "jenkins: COMMAND"

stgraber commented:
What's going on with that build tag commit?
What's it supposed to be doing exactly?

lxd/apparmor doesn't contain any cgo so it makes no sense to have a build-tag requiring it, it also will never be used from within the agent so the agent build tag also makes no sense and yeah, it's linux specific but it's a package that's never included into a non-Linux binary so it makes no sense to use that build tag either.

That's one example, I've seen similar issues with just about all files that are being touched here, so I'm wondering what the motivation was for this change.

stgraber commented:
I haven't reviewed the other commit too closely, mostly because it's very hard in its current state.
You should ideally split it into:
 - lxd/apparmor: Add rsync profile
 - one commit per package using rsync to do the mechanical change needed to use the profile
 - separate commits for anything else that's not a simple mechanical change

Viktor-Yakovchuk commented:
> What's going on with that build tag commit? What's it supposed to be doing exactly?
> 
> lxd/apparmor doesn't contain any cgo so it makes no sense to have a build-tag requiring it, it also will never be used from within the agent so the agent build tag also makes no sense and yeah, it's linux specific but it's a package that's never included into a non-Linux binary so it makes no sense to use that build tag either.
> 
> That's one example, I've seen similar issues with just about all files that are being touched here, so I'm wondering what the motivation was for this change.

I faced up with following problem while compiling lxd with apparmor package was imported in rsync package:
```
CGO_ENABLED=0 go install -v -tags netgo ./lxd-migrate
package github.com/lxc/lxd/lxd-migrate
        imports github.com/lxc/lxd/lxd/rsync
        imports github.com/lxc/lxd/lxd/apparmor
        imports github.com/lxc/lxd/lxd/cgroup
        imports github.com/lxc/lxd/lxd/db/warningtype: build constraints exclude all Go files in /home/ubuntu/lxd/lxd/db/warningtype
make: *** [Makefile:34: build] Error 1
```
To fix it all constraints in first commit added. I also add information about this error in commit message.

Viktor-Yakovchuk commented:
> I haven't reviewed the other commit too closely, mostly because it's very hard in its current state. You should ideally split it into:
> 
>     * lxd/apparmor: Add rsync profile
> 
>     * one commit per package using rsync to do the mechanical change needed to use the profile
> 
>     * separate commits for anything else that's not a simple mechanical change

I have split all changes in separate commits.

stgraber commented:
Okay, we're getting closer to something that will work.
We can't have `lxd/rsync` just rely on AppArmor, that's not going to work as that package is used outside of lxd even on non-Linux environments.

Instead I think what we should do is rework the way rsync is invoked inside of `lxd/rsync` so that we can override it to use AppArmor. Defining a new global variable in `lxd/rsync` called `CommandRunner` or something like that and defined as a `func(args []string, sourcePath string, destPath string) error` or something along those lines.

Then make it so that `lxd/rsync` will work as it does today when that variable isn't set, but when it's set, that function will be called instead of running traditional rsync, allowing us to override it for use within LXD.

This will also mean that `lxd/rsync` never imports `lxd/apparmor` which then avoids all of the build tag issues.

Does that make sense?

Viktor-Yakovchuk commented:
I got rid of the apparmor in rsync and reverted commit with build constraints. Then I faced up with problem while compiling lxd:
```
CGO_ENABLED=0 go install -v -tags netgo ./lxd-migrate
github.com/lxc/lxd/lxd/sys
github.com/lxc/lxd/lxd/rsync
# github.com/lxc/lxd/lxd/rsync
lxd/rsync/rsync.go:24:27: undefined: sys.OS
lxd/rsync/rsync.go:84:27: undefined: sys.OS
lxd/rsync/rsync.go:219:22: undefined: sys.OS
lxd/rsync/rsync.go:288:22: undefined: sys.OS
lxd/rsync/rsync.go:393:34: undefined: sys.OS
```
This appears because of build constraints in `lxd/sys/os.go` file which is `//go:build linux && cgo && !agent`
We can't get rid of sys.OS because it is mandatory for apparmor package.

As an experiment I changed Makefile from this
```
CGO_ENABLED=0 go install -v -tags netgo ./lxd-migrate
```
to this
```
go install -v -tags netgo ./lxd-migrate
```
And had LXD built successfully. 
I also watch through the code of lxd-migrate and found that it only uses `rsync.AtLeast` from rsync package and in my version it stays in file without build constraints. lxd-migrate uses `rsync` entirely without calling any other function form rsync package. Here is an example from `lxd-migrate/transfer.go`:
```
	cmd := exec.CommandContext(ctx, "rsync", args...)
```
So my changes may not interfere with work of lxd-migrate

The question is do we really need to use exactly covered by apparmor rsync calls from anywhere else where new build option prevent us of working with them?

Viktor-Yakovchuk commented:
@stgraber kind reminder

simondeziel commented on lxd/apparmor/rsync.go:
```diff
@@ -0,0 +1,221 @@
+  capability fsetid,
+  capability fowner,
+
+  unix, # TODO Perhaps a stricter restriction can be made here
```
Yeah, it's a bit broad. Could you check what it trips during an run with `flags=(complain)`, please?

Viktor-Yakovchuk replied:
I have added `complain` and have received following log:
```
[Sat Apr 15 13:13:36 2023] audit: type=1400 audit(1681564416.406:56): apparmor="ALLOWED" operation="connect" profile="lxd_rsync-var-snap-lxd-common-lxd-storage-pools-default-containers-tc1" pid=3677 comm="lxd.debug" family="unix" sock_type="stream" protocol=0 requested_mask="send receive connect" denied_mask="send connect" addr=none peer_addr="@lxd/fb3fdc7e-4c97-4b00-b9e3-d920888d83ec" peer="unconfined"
```

I have fixed capability to make it more strict.

Thanks for suggestion.

simondeziel replied:
Thanks, with that information, I'd probably even try with:

```
unix (connect, send, receive) type=stream peer=(addr="@lxd/*"),
```

simondeziel replied:
Another thing probably worth testing is to interrupt an ongoing transfer. This is a situation where LXD would send a signal to rsync requiring an Apparmor rule along those lines (please test and adapt the signal set accordingly):

```
 signal (receive) set=(int,term) peer=(label=unconfined),
```

simondeziel commented on lxd/apparmor/rsync.go:
```diff
@@ -0,0 +1,221 @@
+
+{{- if .dstPath }}
+  {{ .dstPath }}/** rw,
+  {{ .dstPath }}/ rw,
```
I have not tested but would have expected that it would also need `link` as in `rwl,`.

Viktor-Yakovchuk replied:
I have tested it on my clusters. One cluster consists of 3 nodes and uses Ceph. The over is one LXD node, but also uses Ceph and dir as a storage.

To trigger LocalCopy I use:
```
lxc storage create test dir
lxc storage volume create test vl_test1
lxc storage volume snapshot test vl_test1
lxc storage volume restore test vl_test1 snap0
```

To trigger Send and Recv use:
```
lxc copy tc1 cl1:testCopy5
```
where `tc1` is a container and `cl1` is a remote cluster.

simondeziel replied:
Could you try copying a filesystem that contains hardlinks?

simondeziel commented on lxd/apparmor/rsync.go:
```diff
@@ -0,0 +1,221 @@
+  # Snap-specific libraries
+  /snap/lxd/*/lib/**.so* mr,
+
+  /var/snap/lxd/common/lxd.debug mixr,
```
I don't understand why `rsync` would need to execute `lxd.debug`?

Viktor-Yakovchuk replied:
In function `sendSetup` in https://github.com/lxc/lxd/blob/master/lxd/rsync/rsync.go#L124 variable `execPath` has a value of current lxd executable. It is got from https://github.com/lxc/lxd/blob/master/lxd/instance/drivers/driver_lxc.go#L5466 (last argument).

Viktor-Yakovchuk commented on lxd/rsync/rsync.go:
```diff
@@ -149,7 +152,27 @@ func sendSetup(name string, path string, bwlimit string, execPath string, featur
+
+	profileName := apparmor.GetRsyncProfileName(path)
+	newArgs := append([]string{"-p", profileName, "rsync"}, args...)
+	cmd := exec.Command("aa-exec", newArgs...)
```
I have tested it on my clusters. One cluster consists of 3 nodes and uses Ceph. The over is one LXD node, but also uses Ceph and dir as a storage.

To trigger LocalCopy I use:
```
lxc storage create test dir
lxc storage volume create test vl_test1
lxc storage volume snapshot test vl_test1
lxc storage volume restore test vl_test1 snap0
```

To trigger Send and Recv use:
```
lxc copy tc1 cl1:testCopy5
```
where `tc1` is a container and cl1 is a remote cluster.

Viktor-Yakovchuk replied:
I tried to make minimal changes and reusing `shared/subprocess/proc.go` will result more code rewriting.

stgraber commented:
Spent a couple of hours working on this one to use the concepts I described in past comments.
The result is no change to any existing `lxd/rsync` users, no need for extra build tags and overall a much smaller change, including the part done in `lxd/apparmor`.

This seems to be working well enough with the most simple case, `rsync` is called wrapper by AppArmor, it runs, profile gets unloaded properly afterwards and I'm not hitting any DENIED.

Let's see what a full testsuite run does.

stgraber commented:
jenkins: test this please

stgraber commented:
Doing some tweaks to the profile as we're hitting some bits that aren't handled by apparmor.

stgraber commented:
jenkins: test this please

stgraber commented:
jenkins: test this please

stgraber commented:
@tomponline can you give it a quick review? Since I updated this a fair bit, probably not the best idea for me to just hit merge :)

tomponline commented on lxd/rsync/rsync.go:
```diff
@@ -121,7 +121,7 @@ func sendSetup(name string, path string, bwlimit string, execPath string, featur
 	 * hardcoding that at the other end, so we can just ignore it.
 	 */
-	rsyncCmd := fmt.Sprintf("sh -c \"%s netcat %s %s\"", execPath, auds, name)
+	rsyncCmd := fmt.Sprintf("%s netcat %s %s --", execPath, auds, name)
```
The comment above could do with being updated now shell wrapper not used.

tomponline commented on lxd/apparmor/rsync.go:
```diff
@@ -0,0 +1,192 @@
+	}
+
+	revert.Add(func() { _ = unloadProfile(sysOS, profileName, profileName) })
+	revert.Add(func() { _ = deleteProfile(sysOS, profileName, profileName) })
```
deleteProfile calls unloadProfile internally so I dont think we need the call to unloadProfile here.

tomponline commented on lxd/apparmor/rsync.go:
```diff
@@ -0,0 +1,192 @@
+		return "", err
+	}
+
+	revert.Add(func() { os.Remove(profilePath) })
```
Out of interest, can we just always remove this file via a defer once its been loaded successfully or does it need to linger on disk for the duration the profile is used?

stgraber replied:
There used to be a requirement for the profile to exist on-disk for it to be unloaded, so we've usually kept them around until we're done using them.

tomponline reviewed (approved):
@stgraber @Viktor-Yakovchuk thanks! overall this is looking good. I've made a few of comments for minor issues/questions.

stgraber commented on lxd/apparmor/rsync.go:
```diff
@@ -0,0 +1,192 @@
+{{- end }}
+{{- end }}
+
+  deny {{ .rootPath }}/etc/ssl/openssl.cnf r,
```
`deny` rules are just used to silence apparmor, everything is denied by default

stgraber commented:
Feedback adressed.

stgraber commented:
jenkins: test this please

tomponline reviewed (approved):
LGTM!

stgraber commented:
@tomponline should be green enough
